src/components/draw/DrawDirective.js

```diff
@@ -27,889 +28,798 @@ goog.require('ga_permalink');
-                      coords[2] == item[2]) {
-                    coords = item;
-                  } else {
-                    hasUniqueCoords = false;
-                  }
-                });
-                if (hasUniqueCoords) {
-                  return;
+                var lineCoords = geom.getCoordinates()[0];
+
+                if (nbPoint != lineCoords.length) {
+                  // A point is added
+                  nbPoint++;
+                } else {
+                  var firstPoint = lineCoords[0];
+
+                  // We update features and measures
+                  var lastPoint = lineCoords[lineCoords.length - 1];
+                  var lastPoint2 = lineCoords[lineCoords.length - 2];
+
+                  var isSnapOnFirstPoint = (lastPoint[0] == firstPoint[0] &&
+                      lastPoint[1] == firstPoint[1]);
+
+                  // When the last change event is triggered the polygon is
+                  // closed so isSnapOnFirstPoint is true. We need to know
+                  // if on the change event just before, the snap on last
+                  // point was active.
+                  isFinishOnFirstPoint = (!isSnapOnLastPoint &&
+                      isSnapOnFirstPoint);
+
+                  isSnapOnLastPoint = (lastPoint[0] == lastPoint2[0] &&
+                      lastPoint[1] == lastPoint2[1]);
                 }
+                updateHelpTooltip(helpTooltip, tool.id, true,
+                    isFinishOnFirstPoint, isSnapOnLastPoint);
               }
-
-              if (geom instanceof ol.geom.Polygon && !isFinishOnFirstPoint) {
-                // The sketchFeatureArea is automatically closed by the draw
-                // interaction even if the user has finished drawing on the
-                // last point. So we remove the useless coordinates.
-                var lineCoords = featureToAdd.getGeometry().getCoordinates()[0];
-                lineCoords.pop();
-                featureToAdd = new ol.Feature(
-                    new ol.geom.LineString(lineCoords));
-              }
-
-              // Update feature properties
-              if (!featureToAdd.getId() && lastActiveTool) {
-                featureToAdd.setId(lastActiveTool.id + '_' +
-                    new Date().getTime());
-              }
-              if (lastActiveTool) {
-                featureToAdd.set('type', lastActiveTool.id);
+            });
+          }));
+
+          unDrawEvts.push(draw.on('drawend', function(evt) {
+            // Unregister the change event
+            ol.Observable.unByKey(deregFeatureChange);
+            deactivateTool(lastActiveTool);
+            scope.$applyAsync();
+
+            var featureToAdd = evt.feature;
+            var geom = featureToAdd.getGeometry();
+
+            // According to #3319, it seems LineString can be created with one
+            // point (or with an array of exact same points). If it's the case
+            // we ignore it.
```

Comment:
This could be factored into a dedicated function, which would also allow to bail out earlier (as soon as non-unique coords are detected)